R/make-array.R

```diff
@@ -80,6 +80,101 @@ makeMR <- function (subvariables, name, selections, ...) {
     return(vardef)
 }
 
+
+#' Create Multiple Response Variable from Delimited
+#'
+#' Surveys often record multiple response questions in delimited lists where
+#' each respondent's selections are separated by a delimiter like `;` or `|`.
+#' This function breaks the delimited responses into subvariables, uploads those
+#' subvariables to Crunch, and finally creates a multiple response variable from
+#' them.
+#'
+#' @param var The variable containing the delimited responses
+#' @param delim The delimiter separating the responses
+#' @param name The name of the resulting MR variable
+#' @param selected A character string used to indicate a selection
+#' @param not_selected Character string identifying non-selection
+#' @param unanswered Character string indicating non-response
+#' @param ... Other arguments to be passed on to [makeMR()]
+#'
+#' @return
+#' @export
+mrFromDelim <- function(var,
+    delim,
+    name,
+    selected = "selected",
+    not_selected = "not_selected",
+    unanswered = NA,
+    ...) {
+    if (missing(name)) {
+        halt("Must supply a name for the new variable")
+    }
+    if (is.Categorical(var) || is.Text(var)) {
+        v <- as.vector(var)
+    } else {
+        halt(dQuote("var"), " must be a Categorical or Text Crunch Variable.")
+    }
+    uniques <- unique(v[!is.na(v)])
+    cats <- unique(unlist(strsplit(uniques, delim)))
+    vardefs <- lapply(cats, function(x) createSubvarDef(v, x, delim,
+        selected, not_selected, unanswered, missing = is.na(v)))
+    ds <- loadDataset(datasetReference(var))
+    addVariables(ds, vardefs)
+    hide(var)
+    ds <- refresh(ds)
+    return(makeMR(ds[, cats], name = name, selections = selected))
+}
+
+#' createSubvarDef
+#'
+#' This function creates a single subvariable definition based on a character string
+#' to search for and an originating variable. It uses regex to determine whether
+#' a string is present in a delimited list, then substitutes the user supplied values
+#' to indicate selection, non-selection, and missingness.
+#'
+#' TODO: When Crunch allows variable derivation to be created via regex, this should
+#' create a derivation instead of a definition with values.
```

Comment:
Additionally, I just talked with jj, and _it turns out_ zz9 already has a regex option. The zcl function/operator is `~=`. You should be able to create the subvariables derived with that. I did some hacking and managed to create a derived regex variable with:
```
ds$newv <- c("B, C", "C, D", "B", "C, B")

new_cat_type <- list(
    value=list(
        class="categorical",
        categories=list(list("id" = 1, "name" = "B", "numeric_value" = NA, "missing" = FALSE))))
new_cat_ids <- 1
new_cat <- list(column=I(new_cat_ids), type=new_cat_type)

tst_deriv <- zfunc("case", new_cat)
tst_deriv$args[[2]] <- zfunc("~=", ds$newv, 'B')
tst <- VariableDefinition(name = 'test', derivation = tst_deriv)
ds$test <- tst

is.derived(ds$test)
derivation(ds$test)
as.vector(ds$test)
```
I'm sure this could be done in a more concise manner, but at the very least, this would mean we wouldn't have to pull everything down (and also, the MR would update the new MR variable if rows are added). 

Finally, it would be possible to send these derivations as the subvariables of the ultimate MR variable derivation (so that we don't have to add subvariables and then reference them, then hide them, etc. As a first step, I would try to get the architecture you have now (adding separate variables to be used as subvariables) with the zz9/zcl regex, and then coming back to sending them all together after you get that first step.